### PR TITLE
Better Hyundai and Genesis Support, Better future model scaling

### PR DESCRIPTION
**Pull request by emmertex**
This includes https://github.com/commaai/openpilot/pull/481
So should be merged only after #481 is merged

Comments I will place in changes

### CODE REVIEW

File: selfdrive/car/hyundai/carstate.py [hunk 1]
```diff
@@ -16,8 +16,6 @@ def get_can_parser(CP):
 
     ("YAW_RATE", "ESP12", 0),
 
-    ("CF_Gway_DrvSeatBeltInd", "CGW4", 1),
-
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: Unused

File: selfdrive/car/hyundai/carstate.py [hunk 2]
```diff
@@ -60,15 +58,20 @@ def get_can_parser(CP):
 
     ("CF_Lvr_GearInf", "LVR11", 0),        #Transmission Gear (0 = N or P, 1-8 = Fwd, 14 = Rev)
 
-    ("CR_Mdps_DrvTq", "MDPS11", 0),
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: MDPS11 is not active (present, but always 0) on some older Hyundai.
As such, they cannot detect a user touching the wheel.
This message is also in MDPS12, and is present on all vehicles.

File: selfdrive/car/hyundai/carstate.py [hunk 3]
```diff
@@ -79,14 +82,12 @@ def get_can_parser(CP):
   checks = [
     # address, frequency
     ("MDPS12", 50),
-    ("MDPS11", 100),
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: Not used

File: selfdrive/car/hyundai/carstate.py [hunk 4]
```diff
@@ -79,14 +82,12 @@ def get_can_parser(CP):
...
     ("TCS15", 10),
     ("TCS13", 50),
     ("CLU11", 50),
     ("ESP12", 100),
     ("EMS12", 100),
     ("CGW1", 10),
-    ("CGW4", 5),
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: not used

File: selfdrive/car/hyundai/carstate.py [hunk 5]
```diff
@@ -113,7 +115,11 @@ def get_camera_parser(CP):
     ("CF_Lkas_FcwCollisionWarning", "LKAS11", 0),
     ("CF_Lkas_FusionState", "LKAS11", 0),
     ("CF_Lkas_FcwOpt_USM", "LKAS11", 0),
-    ("CF_Lkas_LdwsOpt_USM", "LKAS11", 0)
+    ("CF_Lkas_LdwsOpt_USM", "LKAS11", 0),
+    ("CF_Lkas_Unknown1", "LKAS11", 0),
+    ("CF_Lkas_Unknown2", "LKAS11", 0),
+    ("CF_Lkas_ActToi", "LKAS11", 0),
+    ("CR_Lkas_StrToqReq", "LKAS11", 0)
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: Used for better camera forwarding

File: selfdrive/car/hyundai/carstate.py [hunk 6]
```diff
@@ -166,7 +172,7 @@ def update(self, cp, cp_cam):
     self.v_wheel_fr = cp.vl["WHL_SPD11"]['WHL_SPD_FR'] * CV.KPH_TO_MS
     self.v_wheel_rl = cp.vl["WHL_SPD11"]['WHL_SPD_RL'] * CV.KPH_TO_MS
     self.v_wheel_rr = cp.vl["WHL_SPD11"]['WHL_SPD_RR'] * CV.KPH_TO_MS
-    self.v_wheel = (self.v_wheel_fl + self.v_wheel_fr + self.v_wheel_rl + self.v_wheel_rr) / 4.
+    self.v_wheel = 1.035 * (self.v_wheel_fl + self.v_wheel_fr + self.v_wheel_rl + self.v_wheel_rr) / 4.
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: The stock speed is about 3.8% out, as such, this 3.5% increase makes speedo, speed limit, and cruise speed match.

File: selfdrive/car/hyundai/carstate.py [hunk 7]
```diff
@@ -191,17 +197,16 @@ def update(self, cp, cp_cam):
     self.left_blinker_flash = cp.vl["CGW1"]['CF_Gway_TurnSigLh']
     self.right_blinker_on = cp.vl["CGW1"]['CF_Gway_TSigRHSw']
     self.right_blinker_flash = cp.vl["CGW1"]['CF_Gway_TurnSigRh']
-    self.steer_override = abs(cp.vl["MDPS11"]['CR_Mdps_DrvTq']) > 100.
+    self.steer_override = abs(cp.vl["MDPS12"]['CR_Mdps_StrColTq']) > 1.0
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: StrColTq is a scaled value, so the threshold of touching the wheel also needed changing

File: selfdrive/car/hyundai/carstate.py [hunk 8]
```diff
@@ -234,6 +239,7 @@ def update(self, cp, cp_cam):
     else:
       self.gear_shifter_cluster = "unknown"
 
-    # save the entire LKAS11 and CLU11
+    # save the entire LKAS11, CLU11 and MDPS12
     self.lkas11 = cp_cam.vl["LKAS11"]
     self.clu11 = cp.vl["CLU11"]
+    self.mdps12 = cp.vl["MDPS12"]
```
Comment (emmertex) [inline selfdrive/car/hyundai/carstate.py]: Future PR requirement

File: selfdrive/car/hyundai/hyundaican.py [hunk 9]
```diff
@@ -8,7 +8,7 @@ def make_can_msg(addr, dat, alt):
 
 def create_lkas11(packer, car_fingerprint, apply_steer, steer_req, cnt, enabled, lkas11, hud_alert, keep_stock=False):
   values = {
-    "CF_Lkas_Icon": 3 if enabled else 0,
+    "CF_Lkas_Icon": 2 if (car_fingerprint in LKAS_FEATURES["icon_basic"]) else (3 if enabled else 0),
```
Comment (emmertex) [inline selfdrive/car/hyundai/hyundaican.py]: All Genesis fault when anything other than 2 is sent,

File: selfdrive/car/hyundai/hyundaican.py [hunk 10]
```diff
@@ -26,21 +26,23 @@ def create_lkas11(packer, car_fingerprint, apply_steer, steer_req, cnt, enabled,
     "CF_Lkas_FcwCollisionWarning": lkas11["CF_Lkas_FcwCollisionWarning"] if keep_stock else 0,
     "CF_Lkas_FusionState": lkas11["CF_Lkas_FusionState"] if keep_stock else 0,
     "CF_Lkas_Chksum": 0,
-    "CF_Lkas_FcwOpt_USM": 2 if enabled else 1,
+    "CF_Lkas_FcwOpt_USM": lkas11["CF_Lkas_FcwOpt_USM"] if keep_stock else (2 if enabled else 1),
     "CF_Lkas_LdwsOpt_USM": lkas11["CF_Lkas_LdwsOpt_USM"] if keep_stock else 3,
+    "CF_Lkas_Unknown1": lkas11["CF_Lkas_Unknown1"] if keep_stock else 0,
+    "CF_Lkas_Unknown2": lkas11["CF_Lkas_Unknown2"] if keep_stock else 0,
```
Comment (emmertex) [inline selfdrive/car/hyundai/hyundaican.py]: These unknown messages are used on some models.

**Comment by emmertex** [comment]
This should have driver-torque-reduction modified, due to the different steering wheel torque value.

Can you test on your Santa Fe and suggest the setting you feel suits best?
I would prefer every model use the new message for steering torque, as it is present on all cars, where the one that was in use is not present on all cars.

File: selfdrive/car/hyundai/values.py [hunk 11]
```diff
@@ -43,10 +43,13 @@ class Buttons:
...
-CHECKSUM = {
-  "crc8": [CAR.SANTA_FE],
-  "6B": [CAR.KIA_SORENTO, CAR.GENESIS],
-  "7B": [CAR.KIA_STINGER, CAR.ELANTRA],
+# Lane Keep Assist related Features and Limitations
+LKAS_FEATURES = {
+  "6B": [CAR.KIA_SORENTO, CAR.GENESIS],     # 6 Bytes used in Checksum
+  "7B": [CAR.KIA_STINGER, CAR.ELANTRA],			# 7 Bytes used in Checksum
+  "crc8": [CAR.SANTA_FE],		                # CRC Checksum
+  "icon_basic": [CAR.GENESIS],		          # Anything but 2 for LKAS_Icon causes MDPS Fault
```
Comment (rbiasini) [inline selfdrive/car/hyundai/values.py]: It seems unused. Why not keeping this CHECKSUM specific?

File: selfdrive/car/hyundai/hyundaican.py (hunk 10 above)
Comment (rbiasini) [reply selfdrive/car/hyundai/hyundaican.py]: but these are not defined in the dbc file. Eventually the can packer should return an error (right now it might not, but not by design)

File: selfdrive/car/hyundai/carstate.py (hunk 7 above)
Comment (rbiasini) [reply selfdrive/car/hyundai/carstate.py]: is CR_Mdps_StrColTq the same as CR_Mdps_DrvTq (except for scaling?)

Comment (emmertex) [reply selfdrive/car/hyundai/carstate.py]: Identical
screenshot-20181227183632-988x1166

File: selfdrive/car/hyundai/hyundaican.py (hunk 10 above)
Comment (emmertex) [reply selfdrive/car/hyundai/hyundaican.py]: Thats because I screwed up and sent an old dbc :(
I'll take these out for now, as they have no known effect, just they exist.

File: selfdrive/car/hyundai/values.py (hunk 11 above)
Comment (emmertex) [reply selfdrive/car/hyundai/values.py]: This is used, hyundaican line 11.
As for the soft disable, this usage may need to be in its own PR as it may be .... rejected.
I'll remove it for this PR

Comment (emmertex) [reply selfdrive/car/hyundai/values.py]: Can keep it checksum specific and make one for vehicle limitations seperate if preferred?
I am wishing to remove any vehicle specific code out of everything except values and interface.

File: selfdrive/car/hyundai/carstate.py (hunk 7 above)
Comment (rbiasini) [reply selfdrive/car/hyundai/carstate.py]: cool

File: selfdrive/car/hyundai/values.py (hunk 11 above)
Comment (rbiasini) [reply selfdrive/car/hyundai/values.py]: yeah, icon_basic has nothing to do with the algorithm used for CHECKSUM computation. I would prefer to use 2 different params.

File: selfdrive/car/hyundai/carstate.py [hunk 12]
```diff
@@ -191,17 +195,16 @@ def update(self, cp, cp_cam):
     self.left_blinker_flash = cp.vl["CGW1"]['CF_Gway_TurnSigLh']
     self.right_blinker_on = cp.vl["CGW1"]['CF_Gway_TSigRHSw']
     self.right_blinker_flash = cp.vl["CGW1"]['CF_Gway_TurnSigRh']
-    self.steer_override = abs(cp.vl["MDPS11"]['CR_Mdps_DrvTq']) > 100.
+    self.steer_override = abs(cp.vl["MDPS12"]['CR_Mdps_StrColTq']) > 1.0
     self.steer_state = cp.vl["MDPS12"]['CF_Mdps_ToiActive'] #0 NOT ACTIVE, 1 ACTIVE
     self.steer_error = cp.vl["MDPS12"]['CF_Mdps_ToiUnavail']
     self.brake_error = 0
-    self.steer_torque_driver = cp.vl["MDPS11"]['CR_Mdps_DrvTq']
+    self.steer_torque_driver = cp.vl["MDPS12"]['CR_Mdps_StrColTq']
```
Comment (rbiasini) [inline selfdrive/car/hyundai/carstate.py]: btw, this change requires a change to panda safety as well. I think this should be in its own PR.

File: selfdrive/car/hyundai/carstate.py (hunk 6 above)
Comment (rbiasini) [reply selfdrive/car/hyundai/carstate.py]: ok for now, would be better to put this in `values.py` (see `honda/values.py`)

**Comment by emmertex** [comment]
I'm going to close this and revisit this when I have more time.
